src/rust/lib/graph-editor/src/component/operator.rs

```diff
@@ -0,0 +1,271 @@
+pub trait HtmlComponent {
+    /// Return all `DomSymbol`s that make up this component.
+    fn elements(&self) -> Vec<DomSymbol>;
+
+    /// Change the `DomScene` of all `DomSymbol`s.
+    fn set_layer(&self, scene:&DomScene) {
+        self.elements().iter().for_each(|element| scene.manage(&element))
+    }
+}
+
+/// Provides information and functionality to resize an element. A complex UI component should
+/// implement this and propagate size and layout changes to all its sub-components.
+pub trait Resizable {
+    /// Set the size for the UI component.
+    fn set_size(&self, size:Vector3<f32>);
+    /// Return the size of the UI element.
+    fn size(&self) -> Vector3<f32>;
+}
+
+/// A component that owns a Network. Can be used to create external animations.
+pub trait Networked {
+    /// Return a reference to the components network.
+    fn network(&self) -> &frp::Network;
+}
+
+// ==================================
+// === Fullscreen Operator Handle ===
+// ==================================
+
+/// FullscreenOperatorCellHandle is a helper type that wraps a `FullscreenOperator` and applies an
+/// undos the operator on the inner component. This can be used to ensure that only a single
+/// component is made fullscreen at any time.
+#[derive(Debug,CloneRef,Derivative)]
+#[derivative(Clone(bound=""))]
+#[derivative(Default(bound=""))]
+pub struct FullscreenOperatorHandle<T> {
+    operator: Rc<RefCell<Option<FullscreenOperator<T>>>>
+}
+
+impl<T:display::Object+Resizable+ NativeComponent +CloneRef+Networked+'static> FullscreenOperatorHandle<T> {
+    /// returns whether there is a component that is in fullscreen mode.
+    pub fn is_active(&self) -> bool {
+        self.operator.borrow().is_some()
+    }
+
+    /// Enables fullscreen mode for the given component. If there is another component already in
+    /// fullscreen mode, it disables fullscreen for that component.
+    pub fn set_fullscreen(&self, target:T, scene:Scene) {
+        self.disable_fullscreen();
+        let operator = FullscreenOperator::apply(target,scene);
+        self.operator.set(operator);
+    }
+
+    /// Disables fullscreen mode for the given component.
+    pub fn disable_fullscreen(&self) {
+        if let Some(old) = self.operator.borrow_mut().take() {
+            old.undo();
+        }
+    }
+
+    /// Return a ref clone of the fullscreen element.
+    pub fn get_element(&self) -> Option<T>{
+        self.operator.borrow().as_ref().map(|op| op.target.clone_ref())
+    }
+}
+
+
+
+// ===============================
+// === Layer Management Helper ===
+// ===============================
+
+/// Indicates the required target layer.
+// FIXME this is a layer management hack. Remove this once we have nicer scene layer management.
+#[derive(Debug)]
+pub enum SymbolType {
+    /// A symbol that goes onto the `Main` layer.
+    Main (Symbol),
+    /// A visualisation symbol that goes above the `Main` layer, but below the cursor.
+    Visualisation (Symbol),
+}
```

Comment:
This is confusing. This is meant to be symbol type, but in fact it contains a symbol! Either the name is bad or this should be refactored to separate struct (preferably the later).